File: 0763-partition-labels/0763-partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        
        my_dict = {}
        res = []

        #Make hashy storing last element of each index
        for i in range(len(s)):
            my_dict[s[i]] = i

        print(my_dict)

        #Two pointers
        start = 0
        max_end = 0

        for end in range(len(s)):

            max_end = max(my_dict[s[end]], max_end)
            
            if end == max_end:
                res.append(end - start + 1)
                start = end + 1
        

        return res

        

File: 0763-partition-labels/0763-partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) 
    {

        //Map to keep track of everything we are holding.
        unordered_map<char, int> map;

        for (int i{}; i < s.size(); i++)
        {
            //Overrites the last spot.
            map[s[i]] = i;
        }

        int start = 0;
        int maxEnd = 0;

        for (int i{}; i < s.size(); i++)
        {
            maxEnd = max(maxEnd, map[s[i]]);

            if (i == maxEnd)
            {
                res.push_back(i - start + 1);
                start = i + 1;
            }
        }

        return res;
    }

private:
    vector<int> res;
};